File: backend/app/routers/questions.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select

from ..database import get_session
from ..models import InterviewQuestion, TimelineEvent
from ..schemas import (
    InterviewQuestionCreate,
    InterviewQuestionRead,
    InterviewQuestionUpdate,
)

router = APIRouter(prefix="/api/questions", tags=["questions"])


@router.post("", response_model=InterviewQuestionRead, status_code=201)
def create_question(
    data: InterviewQuestionCreate,
    session: Session = Depends(get_session),
):
    event = session.get(TimelineEvent, data.event_id)
    if event is None:
        raise HTTPException(status_code=404, detail="时间线节点不存在")
    question = InterviewQuestion.model_validate(data)
    session.add(question)
    session.commit()
    session.refresh(question)
    return question


@router.get("", response_model=list[InterviewQuestionRead])
def list_questions(
    event_id: int,
    session: Session = Depends(get_session),
):
    statement = (
        select(InterviewQuestion)
        .where(InterviewQuestion.event_id == event_id)
        .order_by(InterviewQuestion.created_at.asc())
    )
    return session.exec(statement).all()


@router.patch("/{question_id}", response_model=InterviewQuestionRead)
def update_question(
    question_id: int,
    data: InterviewQuestionUpdate,
    session: Session = Depends(get_session),
):
    question = session.get(InterviewQuestion, question_id)
    if question is None:
        raise HTTPException(status_code=404, detail="面试题不存在")
    update_data = data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(question, key, value)
    session.add(question)
    session.commit()
    session.refresh(question)
    return question


@router.delete("/{question_id}", status_code=204)
def delete_question(question_id: int, session: Session = Depends(get_session)):
    question = session.get(InterviewQuestion, question_id)
    if question is None:
        raise HTTPException(status_code=404, detail="面试题不存在")
    session.delete(question)
    session.commit()

File: frontend/src/api/questions.ts
import { apiClient } from './client'

export interface InterviewQuestion {
  id: number
  event_id: number
  question: string
  my_answer: string
  reflection: string
  tags: string
  created_at: string
}

export interface InterviewQuestionCreate {
  event_id: number
  question: string
  my_answer?: string
  reflection?: string
  tags?: string
}

export type InterviewQuestionUpdate = Partial<
  Omit<InterviewQuestion, 'id' | 'event_id' | 'created_at'>
>

export const questionsApi = {
  list: (eventId: number) =>
    apiClient
      .get<InterviewQuestion[]>('/questions', { params: { event_id: eventId } })
      .then(r => r.data),
  create: (data: InterviewQuestionCreate) =>
    apiClient.post<InterviewQuestion>('/questions', data).then(r => r.data),
  update: (id: number, data: InterviewQuestionUpdate) =>
    apiClient.patch<InterviewQuestion>(`/questions/${id}`, data).then(r => r.data),
  delete: (id: number) => apiClient.delete(`/questions/${id}`).then(() => undefined),
}

const NON_INTERVIEW_STAGES = new Set([
  'Offer 沟通',
  '入职',
  '被拒',
  '主动放弃',
  '泡池子',
])

export function isInterviewLikeStage(stage: string): boolean {
  return !NON_INTERVIEW_STAGES.has(stage)
}
